var/spack/repos/builtin/packages/libceed/package.py

```diff
@@ -0,0 +1,62 @@
+##############################################################################
+# Copyright (c) 2013-2016, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the LICENSE file for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from spack import *
+
+import os
+import re
+
+
+class Libceed(Package):
+    """The CEED API Library: Code for Efficient Extensible Discretizations
+    """
+
+    homepage = "https://github.com/CEED/libCEED"
+
+    version('develop', git='https://github.com/CEED/libCEED.git',
+            branch='master')
+    version('0.1', git='https://github.com/CEED/libCEED.git', tag='v0.1')
+
+    variant('occa', default=True, description='Enable OCCA backends.')
+
+    depends_on('occa@develop', when='+occa')
+
+    phases = ['build', 'install']
+
+    def build(self, spec, prefix):
+        make('all')
+
+    def install(self, spec, prefix):
+        make('install', 'DESTDIR=%s' % prefix, parallel=False)
```

Comment:
Should we also copy the `data/` directory as part of `install`?

Without it the examples will be build, but will not be able to run with default parameters...

Reply:
First, I think, we need to add `mfem`, `petsc` and `nek5000` variants of `libceed` that will build their respective examples.

Reply:
@v-dobrev Disagree. Those examples are not installed (and shouldn't be as part of libceed) and don't change the libraries.

Reply:
@jedbrown, I'm talking about variants, i.e. options that users will need to enable if they want them installed. Probably better names for the variants are `examples-mfem`, `examples-petsc`, etc. An alternative is to make a separate package, e.g. `libceed-examples` but I prefer the first option.

Reply:
What is the convention for other packages? Is there an option to install MPICH examples? I don't think this is Spack's business. A user can get the examples anywhere and build them with the libceed that they have installed.

Reply:
It is up to the packages, I presume. I've seen this in `sundials`. Mfem also has two variants: `examples` and `miniapps` - I'm not sure who added them but it seems reasonable - they are part of the repository and the distribution.

Reply:
@tgamblin What's the Spack philosophy about using variants to install examples? No way it would fly in a real distro.

Reply:
We do this for other packages -- it's fine with me.  If other packages happen to need them, they can depend on them explicitly, and if the variant is off by default, the examples won't be installed by default.

Any reason we shouldn't do this?  We try to be a bit more inclusive than mainstream distros.

Reply:
Other packages don't depend on examples. Building a variant means rebuilding the library. The purpose of examples is for a user to read the code and run it, perhaps after editing. They typically can't/shouldn't do that from the install directory so really they copy it to a directory that they own. Variants seems like a blunt instrument to provide the user with yet another convention for finding a place to copy examples from: go find `opt/spack/*/*/libceed-0.1-*/share/libceed/examples/ex1.c` and if that path doesn't exist, re-run spack to install the variant and then look back into that path, but follow whichever hash is newer. Can you imagine typing that non-ironically into your project documentation?

Reply:
It is useful to build the examples in Spack for testing, e.g. MFEM+PETSc examples. 

For libCEED, I am more interested in documenting how to use a Spack install to build the examples in a copy that the users have cloned themselves. This should be closer to the use case of an external application.

Here is one way to do that for MFEM, though I'm not sure that's the best one (@tgamblin and @v-dobrev -- do you have better suggestions?).  What will that look like for libCEED?

### Building the MFEM Examples
The simplest way to use the Spack installation is through the `spack location`.
For example, MFEM-based codes can be simply build as follows:
```console
git clone git@github.com:mfem/mfem.git
cd mfem; git checkout v3.3.2
cd examples
make CONFIG_MK=`spack location --install-dir mfem`/share/mfem/config.mk
cd ../miniapps/electromagnetics
make CONFIG_MK=`spack location --install-dir mfem`/share/mfem/config.mk
```

If that doesn't work for you, an alternative is to first export the Spack
installation to a local directory:
```console
spack view --verbose symlink spack-mfem mfem
```
The `spack-mfem` directory now contains the Spack-build MFEM with all of its
depenencies. In particular, the MFEM library in `spack-mfem/lib` and the
MFEM build configuration file in `spack-mfem/share/mfem/config.mk`.

To build the MFEM examples using this installation, one can do the following:
```console
git clone git@github.com:mfem/mfem.git
cd mfem; git checkout v3.3.2
cd examples/petsc
make CONFIG_MK=../../../spack-mfem/share/mfem/config.mk
cd ..
make CONFIG_MK=../../spack-mfem/share/mfem/config.mk
```

The MFEM miniapps can further be build with:
```console
cd ../miniapps/electromagnetics
make CONFIG_MK=../../../spack-mfem/share/mfem/config.mk
```

Reply:
@tzanio So this seems fine to me and there is no need for a `+examples` variant.